src/umode.rs

```diff
@@ -275,6 +293,89 @@ impl UmodeTask {
         Ok(())
     }
 
+    fn map_guest_range_in_umode_slot<T: GuestStagePagingMode>(
+        vm_pages: FinalizedVmPages<T>,
+        addr: RawAddr<GuestPhys>,
+        len: usize,
+        slot: UmodeSlotId,
+        perm: UmodeSlotPerm,
+    ) -> Result<(RawAddr<SupervisorVirt>, GuestUmodeMapping), Error> {
+        let base = PageAddr::with_round_down(addr, PageSize::Size4k);
+        let end = addr
+            .checked_increment(len as u64)
+            .ok_or(Error::AddressOverflow)?;
+        let umode_mapping = vm_pages
+            .map_in_umode_slot(
+                slot,
+                base,
+                PageSize::num_4k_pages(end.bits() - base.bits()),
+                perm,
+            )
+            .map_err(Error::VmPages)?;
+        let vaddr = umode_mapping
+            .vaddr()
+            .raw()
+            .checked_increment(addr.bits() - base.bits())
+            .ok_or(Error::AddressOverflow)?;
+        Ok((vaddr, umode_mapping))
+    }
+
+    pub fn nop() -> Result<u64, Error> {
+        let ctx = UmodeExecutionContext::<u8> {
+            shared_data: None,
+            req: UmodeRequest::Nop,
+        };
+        Self::execute_request(ctx)
+    }
+
+    pub fn attestation_evidence<T: GuestStagePagingMode>(
+        vm: &FinalizedVm<T>,
+        csr_gpa: RawAddr<GuestPhys>,
+        csr_len: usize,
+        certout_gpa: RawAddr<GuestPhys>,
+        certout_len: usize,
+    ) -> Result<u64, Error> {
+        // Map input CSR in Slot A as read-only.
+        let (csr_vaddr, _csr_mapping) = Self::map_guest_range_in_umode_slot(
+            vm.vm_pages(),
+            csr_gpa,
+            csr_len,
+            UmodeSlotId::A,
+            UmodeSlotPerm::Readonly,
+        )?;
+        // Map output certificate in Slot B as writable.
+        let (certout_vaddr, _certout_mapping) = Self::map_guest_range_in_umode_slot(
+            vm.vm_pages(),
+            certout_gpa,
+            certout_len,
+            UmodeSlotId::B,
+            UmodeSlotPerm::Writable,
+        )?;
+        let attestation_mgr = vm.attestation_mgr();
+        // Gather measurement registers from the attestation manager and transform it in a array.
+        let msmt_genarray = attestation_mgr
+            .measurement_registers()
+            .map_err(Error::Attestation)?;
+        let zero = [0u8; u_mode_api::cert::SHA384_LEN];
+        let mut msmt_regs = [zero; attestation::MSMT_REGISTERS];
+        for (i, r) in msmt_genarray.iter().enumerate() {
+            msmt_regs[i].copy_from_slice(r.as_slice());
+        }
+        // Get the CDI ID for the attestation layer.
+        let cdi_id = [0; 20];
```

Comment:
is there a const for the 20 here?

Reply:
That should be `rice::cdi::CDI_ID_LEN`

Reply:
CDI_ID_LEN, but this is temporary code moved from vm.rs that is removed in the following commits.